File: app/src/main/java/com/theappsolutions/boilerplate/ui/base/BaseFragment.java
package com.theappsolutions.boilerplate.ui.base;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;

import com.annimon.stream.Optional;
import com.theappsolutions.boilerplate.injection.component.FragmentComponent;
import com.theappsolutions.boilerplate.injection.module.FragmentModule;

/**
 * @author Severyn Parkhomenko s.parkhomenko@theappsolutions.com
 * @copyright (c) 2018 TheAppSolutions. (https://theappsolutions.com)
 * <p>
 * Abstract activity that every other Activity in this application must implement. It handles
 * creation of Dagger components and makes sure that instances of ConfigPersistentComponent survive
 * across configuration changes.
 */
public abstract class BaseFragment extends Fragment implements BaseView {

    private Optional<BaseActivity> baseActivity = Optional.empty();
    private Optional<FragmentComponent> fragmentComponent = Optional.empty();

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof BaseActivity) {
            baseActivity = Optional.of((BaseActivity) context);
        }
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        baseActivity.ifPresent(activity ->
                fragmentComponent = Optional.of(activity.activityComponent().fragmentComponent(new FragmentModule(this))));
    }

    @Override
    public void showToastMessage(String text) {
        baseActivity.ifPresent(activity -> activity.showToastMessage(text));
    }

    @Override
    public void showToastMessage(int textRes) {
        baseActivity.ifPresent(activity -> activity.showToastMessage(textRes));
    }

    @Override
    public void goBack() {
        baseActivity.ifPresent(BaseActivity::goBack);
    }

    @Override
    public void close() {
        baseActivity.ifPresent(BaseActivity::close);
    }

    @Override
    public void showProgress() {
        baseActivity.ifPresent(BaseActivity::showProgress);
    }

    @Override
    public void dismissProgress() {
        baseActivity.ifPresent(BaseActivity::dismissProgress);
    }

    public void setToolbarTitle(String title) {
        baseActivity.ifPresent(activity -> activity.setToolbarTitle(title));
    }

    public Optional<FragmentComponent> getFragmentComponent() {
        return fragmentComponent;
    }


    public Optional<BaseActivity> getBaseActivity() {
        return baseActivity;
    }
}

File: app/src/main/java/com/theappsolutions/boilerplate/ui/base/BaseFragment.kt
package com.theappsolutions.boilerplate.ui.base

import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment

import com.annimon.stream.Optional
import com.annimon.stream.function.Consumer
import com.theappsolutions.boilerplate.injection.component.FragmentComponent
import com.theappsolutions.boilerplate.injection.module.FragmentModule

/**
 * @author Severyn Parkhomenko s.parkhomenko@theappsolutions.com
 * @copyright (c) 2018 TheAppSolutions. (https://theappsolutions.com)
 *
 *
 * Abstract activity that every other Activity in this application must implement. It handles
 * creation of Dagger components and makes sure that instances of ConfigPersistentComponent survive
 * across configuration changes.
 */
abstract class BaseFragment : Fragment(), BaseView {

    var baseActivity = Optional.empty<BaseActivity>()
    var fragmentComponent = Optional.empty<FragmentComponent>()

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        if (context is BaseActivity) {
            baseActivity = Optional.of((context as BaseActivity?))
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        baseActivity.ifPresent { activity -> fragmentComponent = Optional.of(activity.activityComponent()?.fragmentComponent(FragmentModule(this))) }
    }

    override fun showToastMessage(text: String) {
        baseActivity.ifPresent { activity -> activity.showToastMessage(text) }
    }

    override fun showToastMessage(textRes: Int) {
        baseActivity.ifPresent { activity -> activity.showToastMessage(textRes) }
    }

    override fun goBack() {
        baseActivity.ifPresent({ it.goBack() })
    }

    override fun close() {
        baseActivity.ifPresent({ it.close() })
    }

    override fun showProgress() {
        baseActivity.ifPresent({ it.showProgress() })
    }

    override fun dismissProgress() {
        baseActivity.ifPresent({ it.dismissProgress() })
    }

    fun setToolbarTitle(title: String) {
        baseActivity.ifPresent { activity -> activity.setToolbarTitle(title) }
    }
}
